proposed/swift-interop.md

```diff
@@ -0,0 +1,183 @@
+
+Rejected Option 4 would be most similar to the Objective-C interop experience. However, this experience would require more work in the JIT/AOT compilers and would make the translation between .NET exception and Swift error codes inflexible. Modern .NET interop solutions generally push error-exception translation mechanisms to be controlled by higher-level interop code generators instead of the runtime for flexibility. Not selecting this option would require all `CallConvSwift` `UnmanagedCallersOnly` methods to wrap their contents in a `try-catch` to translate any exceptions. This is already done for the COM source generator and had to be done by Binding Tools for Swift, so the pattern has a lot of implementation expertise.
+
+##### Async Context Register
+
+In the SwiftAsync calling convention, there is also an Async Context register, similar to the Self and Error registers. Like the error register, the async context must be passed by a pointer value. As a result, similar options apply here, with the same constraints. Additionally, we don't already have an existing `CallConvAsync` calling convention modifier, so going the calling convention route like proposed for the self register is not practical. As a result, we will likely need to use a special type like `SwiftAsyncContext` to represent the async context register, similar to the proposals for the error register.
+
+##### Tuples
+
+In the Swift language, tuples are "unpacked" in the calling convention. Each element of a tuple is passed as a separate parameter. C# has its own language-level tuple feature in the ValueTuple family of types. The Swift/.NET interop story could choose to automatically handle tuple types in a similar way for a `CallConvSwift` function call. However, processing the tuple types in the JIT/AOT compilers is complicated and expensive. It would be much cheaper to defer this work upstack to the language projection, which can split out the tuple into individual parameters to pass to the underlying function. The runtime and libraries teams could add analyzers to detect value-tuples being passed to `CallConvSwift` functions at compile time to help avoid pits of failure. As we expect most developers to use the higher-level tooling and to not use `CallConvSwift` directly, we would likely defer any analyzer work until we have a suitable use case.
```

Comment:
I'm not following the UX considerations here. For Tuples it seems we are okay with projection generation tooling to unpack potentially complex Tuples, but for the Error register asking the projection tooling to insert specific calls to `Marshal.Get/SetLastSwiftError` is too much? How are we reconciling this?

Reply:
The issue with `Marshal.Get/SetLastSwiftError` is more about the cost of reading and writing to thread-local storage than the cost in the projection tooling of inserting the calls.

Reply:
The proposal for `Get/SetLastSwiftError` is to cache the result when coming in from a Swift call/returning from a Swift call like how we handle `Get/SetLastSystemError` in LibraryImport.

Making it directly operate on the register itself would mean that the JITs/AOT compilers/interpreters would need to ensure that the error register is not touched by any other code, user or otherwise from the Swift call to the `GetLastSwiftError` call. Modelling the error as a special type in the signature makes "ensure that we return the value from this register after the call" easier to model.